finesapp/src/main/kotlin/ua/anastasiia/finesapp/controller/grpc/GrpcService.kt

```diff
@@ -0,0 +1,102 @@
+
+import net.devh.boot.grpc.server.service.GrpcService
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.toMono
+import ua.anastasiia.finesapp.ReactorFinesServiceGrpc
+import ua.anastasiia.finesapp.controller.nats.event.NatsEventSubscriber
+import ua.anastasiia.finesapp.dto.response.CarResponse
+import ua.anastasiia.finesapp.dto.response.FineResponse
+import ua.anastasiia.finesapp.dto.response.toCar
+import ua.anastasiia.finesapp.dto.response.toFine
+import ua.anastasiia.finesapp.dto.response.toResponse
+import ua.anastasiia.finesapp.dto.toFine
+import ua.anastasiia.finesapp.dto.toProto
+import ua.anastasiia.finesapp.exception.CarPlateNotFoundException
+import ua.anastasiia.finesapp.exception.CarsNotFoundException
+import ua.anastasiia.finesapp.input.reqreply.car.GetAllCarsRequest
+import ua.anastasiia.finesapp.input.reqreply.car.GetAllCarsResponse
+import ua.anastasiia.finesapp.input.reqreply.fine.GetFineByCarPlateRequest
+import ua.anastasiia.finesapp.input.reqreply.fine.GetFineByCarPlateResponse
+import ua.anastasiia.finesapp.input.reqreply.fine.StreamByCarPlateRequest
+import ua.anastasiia.finesapp.input.reqreply.fine.StreamByCarPlateResponse
+import ua.anastasiia.finesapp.service.FineService
+
+@GrpcService
+class GrpcService(
+    private val fineService: FineService,
+    private val natsEventSubscriber: NatsEventSubscriber
+) : ReactorFinesServiceGrpc.FinesServiceImplBase() {
+
+    override fun getAllCars(request: Mono<GetAllCarsRequest>): Mono<GetAllCarsResponse> {
+        return request
+            .flatMap { fineService.getAllCars().collectList() }
+            .map { cars -> buildAllCarsResponse(cars) }
+            .onErrorResume { exception -> handleCarsException(exception) }
+    }
+
+    override fun getByCarPlate(request: Mono<GetFineByCarPlateRequest>): Mono<GetFineByCarPlateResponse> {
+        return request
+            .flatMap { fineService.getFineByCarPlate(it.carPlate) }
+            .map { fine -> buildGetFineByCarPlateResponse(fine) }
+            .onErrorResume { exception -> handleCarPlateException(exception) }
+    }
+
+    override fun streamByCarPlate(request: Mono<StreamByCarPlateRequest>): Flux<StreamByCarPlateResponse> {
+        return request.flatMapMany { streamByCarPlateRequest ->
+            fineService.getFineByCarPlate(streamByCarPlateRequest.carPlate)
+                .flatMapMany { initFineState ->
+                    natsEventSubscriber.subscribe(streamByCarPlateRequest.carPlate)
+                        .map { trafficTicketAddedEvent ->
+                            buildStreamByCarPlateResponse(trafficTicketAddedEvent.fine.toFine().toResponse())
+                        }.startWith(
+                            buildStreamByCarPlateResponse(initFineState)
+                        )
+                }
+                .onErrorResume { exception -> handleCarPlateStreamException(exception) }
+        }
```

Comment:
NIT
 you combine two flux , but  natsEventSubscriber.subscribe(...) doesnt need t know about initFineState, so you can split these fluxes (maybe later relocate them to private methods)
```
return request.flatMapMany {
    streamByCarPlateRequest ->
    Flux.concat(
        fineService.getFineByCarPlate(streamByCarPlateRequest.carPlate).flatMap { initFineState ->
            buildStreamByCarPlateResponse(initFineState).toMono()
        },
        natsEventSubscriber.subscribe(streamByCarPlateRequest.carPlate)
            .map { trafficTicketAddedEvent ->
                buildStreamByCarPlateResponse(trafficTicketAddedEvent.fine.toFine().toResponse())
            }
    ).onErrorResume { exception -> handleCarPlateStreamException(exception) }
}
```